modules/apigee_edge_apidocs/tests/src/Functional/ApiDocsAccessTest.php

```diff
@@ -0,0 +1,418 @@
+  }
+
+  /**
+   * Set up menus and tasks in their regions.
+   *
+   * Since menus and tasks are now blocks, we're required to explicitly set them
+   * to regions.
+   */
+  protected function setupMenus() {
+    $this->drupalPlaceBlock('system_menu_block:tools', ['region' => 'primary_menu']);
+    $this->drupalPlaceBlock('local_tasks_block', ['region' => 'secondary_menu']);
+    $this->drupalPlaceBlock('local_actions_block', ['region' => 'content']);
+    $this->drupalPlaceBlock('page_title_block', ['region' => 'content']);
+  }
+
+  /**
+   * Test admin access control functionality for apidocs.
+   */
+  public function testApiDocAccessAdmin() {
+    $assert_session = $this->assertSession();
+
+    // Test the 'administer apidoc entities' permission.
+    $this->drupalLogin($this->drupalCreateUser([
+      'administer apidoc entities',
+      'administer apidoc display',
+      'administer apidoc fields',
+      'administer apidoc form display',
+    ]));
+
+    $this->drupalGet('apidoc/' . $this->apidocPublished->id());
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocPublished, 'view', TRUE);
+
+    $this->drupalGet('apidoc/' . $this->apidocUnpublished->id());
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocUnpublished, 'view', TRUE);
+
+    $this->drupalGet('admin/structure/apidoc/' . $this->apidocPublished->id() . '/edit');
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocPublished, 'update', TRUE);
+    $this->drupalGet('admin/structure/apidoc/' . $this->apidocUnpublished->id() . '/edit');
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocUnpublished, 'update', TRUE);
+
+    $this->drupalGet('admin/structure/apidoc/' . $this->apidocPublished->id() . '/delete');
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocPublished, 'delete', TRUE);
+    $this->drupalGet('admin/structure/apidoc/' . $this->apidocUnpublished->id() . '/delete');
+    $assert_session->statusCodeEquals(200);
+    $this->assertApiDocAccess($this->apidocUnpublished, 'delete', TRUE);
+
+    $this->drupalGet('admin/structure/apidoc');
+    $assert_session->statusCodeEquals(200);
+
+    $this->drupalGet('admin/structure/apidoc/add');
+    $assert_session->statusCodeEquals(200);
+
+    $this->drupalGet('admin/config/apidoc/settings');
+    $assert_session->statusCodeEquals(200);
+
+    // Make sure the field manipulation links are available.
+    $assert_session->linkExists('Manage fields');
+    $assert_session->linkExists('Manage form display');
+    $assert_session->linkExists('Manage display');
+  }
+
+  /**
+   * Test no permissions for apidocs.
+   */
+  public function testApiDocAccessNoPermissions() {
+    $assert_session = $this->assertSession();
+
+    $this->drupalLogin($this->drupalCreateUser());
+
+    $this->drupalGet('apidoc/' . $this->apidocPublished->id());
```

Comment:
Do not repeat yourself. Build a matrix from personas (users with XY permission(s)) routes and the expected status codes and iterate over that matrix instead of typing the same path and an expected status code n+1 times in a test.
https://github.com/apigee/apigee-edge-drupal/blob/8.x-1.x/modules/apigee_edge_teams/tests/src/Functional/AccessTest.php#L122

Reply:
I added a TODO to refactor this test as a kernel test. In that, I agree this could be optimized.